src/xkeys.rs

```diff
@@ -0,0 +1,371 @@
+    /// Remember that this value should be treated as a secret. Do not store it for
+    /// any longer than necessary
+    pub fn seed(&self) -> Result<String> {
+        let Some(secret) = &self.secret else {
+            return Err(err!(IncorrectKeyType, "This keypair has no seed"));
+        };
+
+        Ok(encode_seed(&KeyPairType::Curve, &secret.to_bytes()))
+    }
+
+    /// Returns the encoded, human-readable public key of this key pair
+    pub fn public_key(&self) -> String {
+        encode(&KeyPairType::Curve, self.public.as_bytes())
+    }
+
+    pub fn private_key(&self) -> Result<String> {
+        let Some(secret) = &self.secret else {
+            return Err(err!(IncorrectKeyType, "This keypair has no seed"));
+        };
+
+        Ok(encode_prefix(&[PREFIX_BYTE_PRIVATE], &secret.to_bytes()))
+    }
+
+    /// Returns the type of this key pair.
+    pub fn key_pair_type(&self) -> KeyPairType {
+        KeyPairType::Curve
+    }
+
+    pub fn open(&self, input: &[u8], sender: &Self) -> Result<Vec<u8>> {
+        let nonce_size = <SalsaBox as AeadCore>::NonceSize::to_usize();
+
+        let Some(secret_key) = &self.secret else {
+            return Err(err!(SignatureError, "Cannot open without a private key"));
+        };
+
+        if input.len() <= XKEY_VERSION_V1.len() + nonce_size {
+            return Err(err!(InvalidPayload, "Payload too short"));
+        }
+
+        let Some(input) = input.strip_prefix(XKEY_VERSION_V1) else {
+            return Err(err!(InvalidPrefix, "Cannot open message, wrong version"));
+        };
+
+        let (nonce, input) = input.split_at(nonce_size);
+
+        let b = SalsaBox::new(&sender.public, secret_key);
+        b.decrypt(nonce.into(), input)
+            .map_err(|_| err!(InvalidPayload, "Cannot decrypt payload"))
+    }
+
+    /// Seal is compatible with nacl.Box.Seal() and can be used in similar situations for small
+    /// messages. We generate the nonce from crypto rand by default.
+    pub fn seal(&self, input: &[u8], recipient: &Self) -> Result<Vec<u8>> {
+        self.seal_with_rand(input, recipient, &mut rand::rngs::OsRng)
+    }
+
+    pub fn seal_with_rand(
+        &self,
+        input: &[u8],
+        recipient: &Self,
+        rand: impl CryptoRng + RngCore,
+    ) -> Result<Vec<u8>> {
+        let Some(private_key) = &self.secret else {
+            return Err(err!(SignatureError, "Cannot seal without a private key"));
+        };
+
+        let b = SalsaBox::new(&recipient.public, private_key);
+        let nonce = SalsaBox::generate_nonce(rand);
+        let crypted = b.encrypt(&nonce, input).unwrap(); // Can't fail
+
+        let mut out = Vec::with_capacity(
+            XKEY_VERSION_V1.len()
+                + <SalsaBox as AeadCore>::NonceSize::to_usize()
+                + input.len()
+                + <SalsaBox as AeadCore>::TagSize::to_usize(),
+        );
+        out.extend_from_slice(XKEY_VERSION_V1);
+        out.extend_from_slice(nonce.as_slice());
```

Comment:
This can likely use `extend` as well. It looks like  the `GenericArray` type `generate_nonce` returns implements `IntoIterator` so you should be able to pass it in directly